Associate debug Ranges with MethodEntries

zakkak opened the pull request:
As discussed in https://github.com/oracle/graal/pull/3290 this PR ensures that every debug `Range` is associated with a `MethodEntry`. This PR will enable us to continue working on https://github.com/oracle/graal/pull/2880 and associate the debug info for nested inline methods with `MethodEntry` instances.

zakkak commented:
@adinn please review

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:
```diff
@@ -335,10 +337,9 @@ public ClassEntry getSuperClass() {
+        FileEntry fileEntryToUse = method.fileEntry;
         if (fileEntryToUse == null) {
             /*
              * Search for a matching method to supply the file entry or failing that use the one
```
So this iteration through the methods is now no longer needed because you have already done it under ensureMethodEntry.

zakkak replied:
Correct. My plan is to fix this with https://github.com/oracle/graal/pull/3290

adinn replied:
Well, you should really delete the loop and comment as part of this commit since your existign changes have already made it redundant.

zakkak replied:
OK, done in https://github.com/oracle/graal/pull/3304/commits/55503f9cc20fe2a80ddbe5adee0f839450eea777

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DebugInfoBase.java:
```diff
@@ -236,40 +236,38 @@ public void installDebugInfo(DebugInfoProvider debugInfoProvider) {
             int lo = debugCodeInfo.addressLo();
             int hi = debugCodeInfo.addressHi();
             int primaryLine = debugCodeInfo.line();
             boolean isDeoptTarget = debugCodeInfo.isDeoptTarget();
```
I think we probably ought to migrate DebugCodeInfo::isDeoptTarget() to DebugMethodInfo and cache this value in the MethodEntry when we create it. The test simply checks that the hosted method name ends in "**" and coudl be done up front when processing the DebugMethodInfo. If we do that then there is no need to retrieve isDeoptTarget here and pass it into makePrimaryRange. The latter function can read it off the methodEntry.

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:
```diff
@@ -356,6 +357,19 @@ public Range makePrimaryRange(String methodName, String symbolName, String param
+    }
+
+    public MethodEntry ensureMethodEntry(DebugInfoProvider.DebugMethodInfo debugMethodInfo, DebugInfoBase debugInfoBase, DebugContext debugContext) {
+        String methodName = debugInfoBase.uniqueDebugString(debugMethodInfo.name());
```
Import DebugInfoProvider.DebugMethodInfo so you don't have to have a qualified type in the argument list.

zakkak replied:
The import was actually already present :/
Fixed.

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java:
```diff
@@ -29,13 +29,18 @@
 public class MethodEntry extends MemberEntry {
     TypeEntry[] paramTypes;
     String[] paramNames;
+    String[] paramTypeNames;
```
I'm not sure why you are materialising array paramTypeNames here rather than retrieve the names from the respective element of paramTypes. It seems an unnecessary waste of space to avoid a single field indirection. Or is there a good reason for this chnage?

adinn replied:
Ah ok, I see this is because you need to use it to implement Range::getParamTypeNames. Ok.

zakkak replied:
Me neither, I think at the point of adding it I was using it in more places. Removed.

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java:
```diff
@@ -724,18 +724,13 @@ private int writeMethodParameterDeclarations(DebugContext context, ClassEntry cl
+            String paramName = uniqueDebugString("");
+            FileEntry fileEntry = range.getFileEntry();
+            if (fileEntry != null) {
+                pos = writeMethodParameterDeclaration(context, paramName, paramTypeName, false, isSpecification, buffer, pos);
```
Hmm, this if/else sequence is wrong -- but not because of anything you have done. Passing paramTypeName in place of paramName in the else branch makes no sense at all (paramName is ignored for now). The method call here should include a filename argument or null. This argument should drive the choice of abbrev code as either DW_ABBREV_CODE_method_parameter_declaration2 or DW_ABBREV_CODE_method_parameter_declaration3. It does not matter too much as we don't have a line or column number so a file name for the parameter declaration is not a lot of use. Still it needs fixing, especially as the current code makes no sense at all. I'll raise a separate issue for this.

adinn commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -195,6 +196,17 @@ protected static ResolvedJavaType getJavaType(HostedType hostedType, boolean wan
         return hostedType.getWrapped().getWrapped();
     }
 
+    protected static ResolvedJavaType getJavaType(AnalysisType analysisType, boolean wantOriginal) {
```
Why do you need this? Where are you seeing Analysis types pop up? I think you ought only to be seeing HostedTypes.

zakkak replied:
Hmmm, you are right. I guess I somehow thought I should include it like we do in `computeFullFilePath` (but there it's just a step after unwrapping a `HostedType`). Anyhow this is now removed.

zakkak commented:
Thanks for the review @adinn . I have now resolved the issues you pointed out.

adinn reviewed (commented):
This is looking good apart from the eclipse style check failure.

zakkak commented:
@olpaw could you please review this?

adinn commented:
@olpaw Any chance of this getting a review from you or someone else who can get it through the gate. I believe the latest changes are sound.

adinn commented:
@olpaw Any chance of this getting a review from you or someone else who can get it through the gate. I believe the latest changes are sound.

zakkak commented:
This is now rebased (cleanly) to latest `master`.

@adinn https://github.com/oracle/graal/pull/3304/commits/acfec7abc76e2d14b3a0cf576f2845d6f1dcad07 and https://github.com/oracle/graal/pull/3304/commits/753a5951a766ffd6a7820125ed2826b0d5c384be are new in case you want to have a look at them.

adinn commented:
I'm not happy with the change to remove the method notifications during presentation of the type info.

Of course, method details presented by NativeImageDebugInstanceTypeInfo::methodInfoProvider and method range details presented later by NativeImageDebugInfo::codeInfoProvider ought to match up 1-1. You should never see range info for a method that has not been advertised using exactly the same classname, method name and signature as belonging to a type and vice versa. However, that's not a reason to drop one stream of data in favour of the other.

The two streams of data are produced from two independent data sets created by the GraalVM image generator which are then transformed by the DebugInfoProvider code to report them using a consistent mapping of HostedXXX objects to their String identifier names. The data sets and the  mapping need to be checked for consistency. It took some time to ensure the right names were used and it would be very easy for a change in the closed world analysis or compilation phase to inadvertently break it.

If you are worried about the performance issues when it comes to linking range info to method info then you can always use sorting and/or hashing to speed up mapping a range to a method. The same approach should also make it quicker to verify the reverse constraint that all methods are accounted for. If that verify step is too expensive then do it in a method called under an assert so it will not get run in production.

zakkak commented:
Following @adinn 's recommendation https://github.com/oracle/graal/commit/753a5951a766ffd6a7820125ed2826b0d5c384be is now replaced by https://github.com/oracle/graal/pull/3304/commits/40b5e8d7155814ae152964d9ef84029a8eb8a026

zakkak commented:
CI fails with:

```
There was 1 failure:
1) com.oracle.truffle.api.test.polyglot.PolyglotCachingTest#testSourceFreeContextStrong
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.oracle.truffle.api.test.GCUtils.assertObjectsCollectible(GCUtils.java:88)
	at com.oracle.truffle.api.test.polyglot.PolyglotCachingTest.testSourceFreeContextStrong(PolyglotCachingTest.java:165)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at com.oracle.mxtool.junit.MxJUnitWrapper$1.schedule(MxJUnitWrapper.java:324)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:357)
	at com.oracle.mxtool.junit.MxJUnitWrapper.main(MxJUnitWrapper.java:222)

FAILURES!!!
Tests run: 3983,  Failures: 1
```

I do not see how the changes of this PR could be causing this :/. I think that the failure is irrelevant to the PR, if not please let me know.

adinn commented:
Hi Foivis,

I am still not clear this is correct. I said this earlier
> method details presented by NativeImageDebugInstanceTypeInfo::methodInfoProvider and method range details presented later by NativeImageDebugInfo::codeInfoProvider ought to match up 1-1

When I said method range I was talking about the *primary* ranges. Your patch seems to interpret my comment as suggesting that every *subrange* whose details are presented under NativeImageDebugInfo::codeInfoProvider will match up with a method presented via NativeImageDebugInstanceTypeInfo::methodInfoProvider. I don't believe that is true. Indeed, your experience with prior inline patches suggests to me that it is false.

So, in class DebugInfoBase at line 266 you have this call

    266             MethodEntry subMethodEntry = subClassEntry.getMethodEntry(debugLineInfo, this);

In the case where the method does not exist you will hit the assert in ClassEntry at line 365

    365         assert false : "Method " + returnTypeName + " " + methodName + "(" + paramSignature + ") not found in ClassEntry of " + getTypeName();

That is not going to be good enough.

In this situation I believe you will need to create a dummy MethodEntry to represent a method that only appears inline. Also, if there is no ClassEntry for the class named in the method details then you will need to create a dummy ClassEntry for a class whose methods only ever appear inline.

These will be needed later on when you want to generate inline method info as they will define the class and method to which the inline details correspond. These dummy entries will need to have DWARF info generated for them but it will be a reduced set compared to the DWARF record we see for classes and methods that are notified via NativeImageDebugInstanceTypeInfo.

zakkak commented:
> When I said method range I was talking about the _primary_ ranges. Your patch seems to interpret my comment as suggesting that every _subrange_ whose details are presented under NativeImageDebugInfo::codeInfoProvider will match up with a method presented via NativeImageDebugInstanceTypeInfo::methodInfoProvider. I don't believe that is true. 

You have mentioned this concern before in https://github.com/oracle/graal/pull/2880#discussion_r596058717 but I still believe that all the info we need are provided by `NativeImageDebugInstanceTypeInfo::methodInfoProvider`.

> Indeed, your experience with prior inline patches suggests to me that it is false.

Can you please point me to such an example? 
I recall telling you (in a chat) that I hit such a case, but if I am not mistaken that turned out to be due to a different issue.

In any case I prefer this patch going in as is, if you agree that it doesn't break anything at the moment, and we can revisit the _assertion vs insert_ in #2880 where we will be able to verify the assertions by running tests with inlined methods and assertions enabled.

adinn commented:
@zakkak The problem I have here is that you want to modify this to something that as far as I can tell might break.

That's not just 'might break' as in we'll find out soon enough and fix it. That's 'might break' as in we could release this and only later find out that some users are unable to include debug info because their apps invoke inlined methods which 1) have not been notified by NativeImageDebugInstanceTypeInfo::methodInfoProvideror or 2) belong to a class that has not been notified by NativeImageDebugInstanceTypeInfo::methodInfoProvidertypeInfoProvider. If that happens for an app deployed using this release that needs to be debugged to work out why it is broken then we have a problem.

You say

> we will be able to verify the assertions by running tests with inlined methods and assertions enabled

At the very least we would need to run a comprehensive suite of tests with many different apps (and probably kinds of app) before we can safely assume that. What have you tested this on?

Essentially, I need a good reason to trust this problem is not going to happen. I have studied the Analysis code but I cannot say for sure that whenever an inlined method appears in the CompiledCode a corresponding class and method will have been notified via the DebugTypeInfo interface. I would be happy to accept your patch if an Oracle GraalVM dev could confirm that you are right. Equally if you could explain to me how the Analysis and compilation processes engineer this situation. Otherwise I think you need to allow for missing methods and types.

zakkak commented:
> At the very least we would need to run a comprehensive suite of tests with many different apps (and probably kinds of app) before we can safely assume that. What have you tested this on?

My primary driver is Quarkus' Integration Tests, but for testing inlined methods I also plan to add explicitly inlined methods to our debug info test.

> Essentially, I need a good reason to trust this problem is not going to happen. I have studied the Analysis code but I cannot say for sure that whenever an inlined method appears in the CompiledCode a corresponding class and method will have been notified via the DebugTypeInfo interface. I would be happy to accept your patch if an Oracle GraalVM dev could confirm that you are right. Equally if you could explain to me how the Analysis and compilation processes engineer this situation. Otherwise I think you need to allow for missing methods and types.

ACK, since that would take me some time (and I am not sure when I will be able to allocate that time :) ) I'll bring back the "insertion" code (unless someone from Oracle happen to confirm my claims before doint it) and come back with a more convincing story :).

zakkak commented:
I rebased the branch and brought back the code that creates a new MethodEntry in case it fails to find one.
@adinn FYI the new changes are in  https://github.com/oracle/graal/pull/3304/commits/d708a1fc2f81a6ea77c57cefcfac5673fcbd9fc9 , you have already reviewed the rest of the commits (the rebase did not have any conflicts)

adinn reviewed (approved):
Ok, this looks good now.

olpaw commented:
> Ok, this looks good now.

@adinn, @zakkak, sorry I lost track of what was going on with this PR. I will take a look today and tomorrow.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:
```diff
@@ -294,7 +294,9 @@ protected void processMethod(DebugMethodInfo debugMethodInfo, DebugInfoBase debu
          */
         FileEntry methodFileEntry = debugInfoBase.ensureFileEntry(fileName, filePath, cachePath);
-        methods.add(new MethodEntry(methodFileEntry, methodName, this, resultType, paramTypeArray, paramNameArray, modifiers));
+        final MethodEntry methodEntry = new MethodEntry(methodFileEntry, methodName, this, resultType, paramTypeArray, paramNameArray, modifiers);
```
We generally do use `final` only for fields (on locals it's mostly verbose). But you can keep it if you like.

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -958,11 +981,33 @@ public int getFrameSize() {
+            LinkedList<String> paramTypes = new LinkedList<>();
+            Signature signature = hostedMethod.getSignature();
+            for (int i = 0; i < signature.getParameterCount(false); i++) {
+                final JavaType parameterType = signature.getParameterType(i, null);
```
See https://github.com/oracle/graal/pull/3304#discussion_r627286481

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -958,11 +976,33 @@ public int getFrameSize() {
+
+        @Override
+        public List<String> paramTypes() {
+            LinkedList<String> paramTypes = new LinkedList<>();
```
What's the point of using linked list (here and elsewhere). Using linked list is almost always a bad idea.

zakkak replied:
There's no excuse for it... Thanks for catching it!

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -958,11 +976,33 @@ public int getFrameSize() {
         @Override
-        public int getModifiers() {
+        public List<String> paramNames() {
+            /* Can only provide blank names for now. */
```
How about
```java
String[] paramNames = new String[signature.getParameterCount(false)] ;
Arrays.fill(paramNames, "");
return Arrays.asList(paramNames);
```

zakkak replied:
Sure, but at some point `paramNames` will not be `""`.

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -1057,11 +1106,26 @@ public String methodName() {
 
+        @Override
+        public boolean isDeoptTarget() {
+            return name().endsWith(HostedMethod.METHOD_NAME_DEOPT_SUFFIX);
```
How about `hostedMethod.compilationInfo.isDeoptTarget()` instead?

zakkak replied:
@adinn this was copied from your implementation, do you have any objections on changing it to `hostedMethod.isDeoptTarget()`?

adinn replied:
> do you have any objections on changing it to hostedMethod.isDeoptTarget()

Not at all. I was not aware that such a method existed when I coded my version. It is far better to use the API method call than rely on the internal details of how deopt method names are encoded.

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -1081,6 +1145,33 @@ public int line() {
+            LinkedList<String> paramNames = new LinkedList<>();
+            Signature signature = method.getSignature();
+            for (int i = 0; i < signature.getParameterCount(false); i++) {
+                paramNames.add("");
```
See https://github.com/oracle/graal/pull/3304/files#r627316683

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:
```diff
@@ -335,27 +338,52 @@ public ClassEntry getSuperClass() {
+        return newMethodEntry;
+    }
+
+    private boolean listIsSorted(List<MethodEntry> list) {
```
Since this just for assertion checking how about:
```java
list.stream().sorted().collect(Collectors.toList()).equals(list);
```

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java:
```diff
@@ -72,26 +74,51 @@ public String getParamName(int idx) {
-        if (!methodName.equals(this.memberName)) {
-            return false;
+    public int compareTo(String methodName, String paramSignature, String returnTypeName) {
+        if (!memberName.equals(methodName)) {
```
I would build this up as
```java
int memberNameCompare = memberName.compareTo(methodName);
if (memberNameCompare != 0) {
    return memberNameCompare;
}
int valueTypeCompare = valueType.getTypeName().compareTo(returnTypeName);
if (valueTypeCompare != 0) {
... etc etc ...
```

zakkak replied:
Refactored, but as:
```
int result = memberName.compareTo(methodName);
if (result == 0) {
    result = valueTypeCompare = valueType.getTypeName().compareTo(returnTypeName);
}
if (result == 0) {
    ... etc etc ...
return result;
```

instead of as suggested. I find this equally readable while a bit more compact.

olpaw replied:
Hmm, so if the first assignment to `result` returns a value `!=` `0` you will have have a couple more `if`s to go though before finally returning that value assigned in the beginning although it was clear from the beginning that this is the value that was going to be returned? :thinking:

olpaw replied:
Of course the compiled version is likely to have optimized early exits. But while it runs in the interpreter it does exactly what you ask it to do. Also I find it easier to reason about code where locals are all effectively final (which is not the case in your variant)

Anyway, if you prefer this style I won't prevent you from merging it like that. I'm just providing my 2-cents.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java:
```diff
@@ -72,26 +74,51 @@ public String getParamName(int idx) {
+    @Override
+    public int compareTo(MethodEntry other) {
+        assert other != null;
+        int result = methodName().compareTo(other.methodName());
```
This could also profit from the strategy shown in https://github.com/oracle/graal/pull/3304/files#r627333745

zakkak replied:
I personally prefer this over https://github.com/oracle/graal/pull/3304/files#r627333745

olpaw reviewed (changes requested):
See comments. Please also remove final from locals where it is obvious that the local is only assigned once.

zakkak commented:
Thanks for the review @olpaw ! I have addressed most of your comments (for the rest please see my replies).

Could you please also provide some info regarding the following comment?

> Your patch seems to interpret my comment as suggesting that every subrange whose details are presented under `NativeImageDebugInfo::codeInfoProvider` will match up with a method presented via `NativeImageDebugInstanceTypeInfo::methodInfoProvider`. I don't believe that is true.

Source: https://github.com/oracle/graal/pull/3304#issuecomment-816780087

The `NativeImageDebugInstanceTypeInfo::methodInfoProvider` essentially gets all types present in the "universe" through `heap.getUniverse().getTypes()` and for each one of them it then gets the declaredMethods through `hostedType.getAllDeclaredMethods()`.  While `NativeImageDebugInfo::codeInfoProvider` gets all the compilations through `codeCache.compilations.entrySet()` and for each one of them it then gets the methods embedded in it through `compilation.getSourceMappings()`.

To my understanding the set of methods returned by the `methodInfoProvider` is a superset of the set of methods returned by the `codeInfoProvider` but @adinn is skeptical about it. We would really appreciate it if you could help us understand what's the case (and why if possible).

zakkak commented:
@olpaw this is rebased to master and now resolves all your comments (except the one about the empty parameter names).
The CI failure seems unrelated to the PR.
It should be ready for merge :)

olpaw commented:
@zakkak I just saw that there other places in the debuginfo sources that use `LinkedList` (e.g. `com.oracle.objectfile.debugentry.ClassEntry#ClassEntry`). Please go over it and replace them with `ArrayList` whereever possible.

olpaw commented:
> I will do some testing to see if I can find examples to support or disprove this theory that I can present tomorrow.

@zakkak if you are right then `com.oracle.svm.hosted.image.NativeImageDebugInfoProvider.NativeImageDebugLineInfo#ownerType` should always successfully cast to `HostedMethod` and the else case should actually be an assertion. Do you agree?
```java
@Override
public String ownerType() {
    if (method instanceof HostedMethod) {
        return getJavaType((HostedMethod) method, true).toJavaName();
    } else {
        assert false: "NodeSourcePosition.getMethod() should always be a HostedMethod"; // <<<<<<<<<<
        return method.getDeclaringClass().toJavaName();
    }
}
```

zakkak commented:
> @zakkak I just saw that there are other places in the debuginfo sources that use `LinkedList` (e.g. `com.oracle.objectfile.debugentry.ClassEntry#ClassEntry`). Please go over it and replace them with `ArrayList` whereever possible.

Resolved in https://github.com/oracle/graal/pull/3304/commits/8ed65a8bd1d6eab42d18b0e9c501b71183e5e3ac

> Actually assuming to be always a `HostedMethod` is incorrect. But assuming it to be `com.oracle.svm.core.meta.SharedMethod` would be correct (`SharedMethod` is what you see if you have Graal running at image-runtime).

I am not sure TBH. The way I tested my hypothesis was by adding an assertion in `com.oracle.objectfile.debugentry.ClassEntry#getMethodEntry` where we essentially search for a method entry (in the method entry set provided by the `methodInfoProvider`) while parsing the data provided by the `codeInfoProvider`. If the search fails and we need to add a new method (`com/oracle/objectfile/debugentry/ClassEntry.java:370`) then the assert should hit and disprove my hypothesis. My testing so far has failed to hit that assert, but I am not confident about the coverage of my testing :)

olpaw reviewed (changes requested):
See https://github.com/oracle/graal/pull/3304#discussion_r629947799

olpaw commented:
`mx eclipseformat` found another style issue. I will fix that on my side.

zakkak commented:
> That said, assuming that you will find all HostedMethods that exist after aot-compilation only by iterating the hosted universe (heap.getUniverse().getTypes() and then going though each type with hostedType.getAllDeclaredMethods()) is not valid. E.g. method specialization can result in having multiple versions of a HostedMethod where you would only see a single one if you just iterated through the hosted universe. Debuginfo generation needs to be able to handle that. E.g. method specialization can result in having multiple versions of a HostedMethod where you would only see a single one if you just iterated through the hosted universe.

That makes sense, but I think it doesn't answer the question we are interested in, possibly because I didn't ask properly so please let me rephrase.

Our goal here is to be able to map an address range to a debuginfo method-declaration entry where the debuginfo method-declaration entry essentially contains info about where (in the source code) the method is defined, what its type is, etc.    
The `SourceMapping`s help us do this by mapping address ranges (`sourceMapping.getStartOffset()`--`sourceMapping.getEndOffset()`) to `ResolvedJavaMethod`s. Then, through the `ResolvedJavaMethod`s we are able to get access to the corresponding source Java method and construct a debug method entry.

The question is if there is any chance to find a Java method through the `SourceMapping`s that we have not seen before through the `methodInfoProvider`. 

More practically, do you think there is any chance to reach https://github.com/zakkak/mandrel/blob/0ea21026b4507d2d29908093b548a9e47361347b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java#L370 since we create method entries for each method of each class in the "universe" in https://github.com/zakkak/mandrel/blob/0ea21026b4507d2d29908093b548a9e47361347b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java#L141 ?

> method specialization can result in having multiple versions of a HostedMethod where you would only see a single one if you just iterated through the hosted universe

This single one though should be able to provide me with enough info to generate the method-declaration entry, right?

> `mx eclipseformat` found another style issue. I will fix that on my side.

Oh the comment is too long...
Thanks for taking care of it.

olpaw commented:
> More practically, do you think there is any chance to reach https://github.com/zakkak/mandrel/blob/0ea21026b4507d2d29908093b548a9e47361347b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java#L370 since we create method entries for each method of each class in the "universe" in https://github.com/zakkak/mandrel/blob/0ea21026b4507d2d29908093b548a9e47361347b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java#L141 ?

Right, unless something dodgy is going on in the backend you should not hit the `newMethodEntry`-part of `com.oracle.objectfile.debugentry.ClassEntry#getMethodEntry`.

> 
> > method specialization can result in having multiple versions of a HostedMethod where you would only see a single one if you just iterated through the hosted universe
> 
> This single one though should be able to provide me with enough info to generate the method-declaration entry, right?

Yes.

olpaw commented:
@zakkak all gates but one are passing internally. And that failing gate has nothing to do with your PR. We are working on getting this sorted. In other words ... it might take a few days before you see your changes showing up on master.

zakkak commented:
> @zakkak all gates but one are passing internally. And that failing gate has nothing to do with your PR. We are working on getting this sorted. In other words ... it might take a few days before you see your changes showing up on master.

OK, thanks for the update Paul.
